File: E_Kotlin/Projects/JuegoSnake/app/src/main/java/com/sebss/juegosnake/androidimpl/SingleTouchHandler.java
package com.sebss.juegosnake.androidimpl;

import java.util.ArrayList;
import java.util.List;

import android.annotation.SuppressLint;
import android.view.MotionEvent;
import android.view.View;

import com.sebss.juegosnake.Pool;
import com.sebss.juegosnake.Input.TouchEvent;
import com.sebss.juegosnake.Pool.PoolObjectFactory;

public class SingleTouchHandler implements TouchHandler {
    boolean isTouched;
    int touchX;
    int touchY;
    Pool<TouchEvent> touchEventPool;
    List<TouchEvent> touchEvents = new ArrayList<>();
    List<TouchEvent> touchEventsBuffer = new ArrayList<>();
    float scaleX;
    float scaleY;

    public SingleTouchHandler(View view, float scaleX, float scaleY) {
        PoolObjectFactory<TouchEvent> factory = TouchEvent::new;
        touchEventPool = new Pool<>(factory, 100);
        view.setOnTouchListener(this);

        this.scaleX = scaleX;
        this.scaleY = scaleY;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        synchronized(this) {
            TouchEvent touchEvent = touchEventPool.newObject();
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    touchEvent.type = TouchEvent.TOUCH_DOWN;
                    isTouched = true;
                    break;
                case MotionEvent.ACTION_MOVE:
                    touchEvent.type = TouchEvent.TOUCH_DRAGGED;
                    isTouched = true;
                    break;
                case MotionEvent.ACTION_CANCEL:
                case MotionEvent.ACTION_UP:
                    touchEvent.type = TouchEvent.TOUCH_UP;
                    isTouched = false;
                    break;
            }

            touchEvent.x = touchX = (int)(event.getX() * scaleX);
            touchEvent.y = touchY = (int)(event.getY() * scaleY);
            touchEventsBuffer.add(touchEvent);

            return true;
        }
    }

    @Override
    public boolean isTouchDown(int pointer) {
        synchronized(this) {
            if(pointer == 0)
                return isTouched;
            else
                return false;
        }
    }

    @Override
    public int getTouchX(int pointer) {
        synchronized(this) {
            return touchX;
        }
    }

    @Override
    public int getTouchY(int pointer) {
        synchronized(this) {
            return touchY;
        }
    }

    @Override
    public List<TouchEvent> getTouchEvents() {
        synchronized(this) {
            int len = touchEvents.size();
            for( int i = 0; i < len; i++ )
                touchEventPool.free(touchEvents.get(i));
            touchEvents.clear();
            touchEvents.addAll(touchEventsBuffer);
            touchEventsBuffer.clear();
            return touchEvents;
        }
    }
}

File: E_Kotlin/Projects/3DGraphicsMotor/app/src/main/java/com/sebss/a3dgraphicsmotor/androidimpl/SingleTouchHandler.kt
package com.sebss.a3dgraphicsmotor.androidimpl

import android.annotation.SuppressLint
import android.view.MotionEvent
import android.view.View
import com.sebss.a3dgraphicsmotor.interfaces.Input
import com.sebss.a3dgraphicsmotor.interfaces.Pool
import com.sebss.a3dgraphicsmotor.interfaces.Pool.PoolObjectFactory

class SingleTouchHandler(view: View, scaleX: Float, scaleY: Float) : TouchHandler {
    private var isTouched = false
    private var touchX = 0
    private var touchY = 0
    private var touchEventPool: Pool<Input.TouchEvent>
    private val touchEvents: MutableList<Input.TouchEvent> = ArrayList()
    private var touchEventsBuffer: MutableList<Input.TouchEvent> = ArrayList()
    private var scaleX: Float
    private var scaleY: Float

    init {
        val factory: PoolObjectFactory<Input.TouchEvent> =
            object : PoolObjectFactory<Input.TouchEvent> {
                override fun createObject(): Input.TouchEvent {
                    return Input.TouchEvent()
                }
            }
        touchEventPool = Pool(factory, 100)
        view.setOnTouchListener(this)
        this.scaleX = scaleX
        this.scaleY = scaleY
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        synchronized(this) {
            val touchEvent = touchEventPool.newObject()
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_DOWN
                    isTouched = true
                }

                MotionEvent.ACTION_MOVE -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_DRAGGED
                    isTouched = true
                }

                MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_UP
                    isTouched = false
                }
            }
            touchX = (event.x * scaleX).toInt()
            touchEvent.x = touchX
            touchY = (event.y * scaleY).toInt()
            touchEvent.y = touchY
            touchEventsBuffer.add(touchEvent)
            return true
        }
    }

    override fun isTouchDown(pointer: Int): Boolean {
        synchronized(this) { return if (pointer == 0) isTouched else false }
    }

    override fun getTouchX(pointer: Int): Int {
        synchronized(this) { return touchX }
    }

    override fun getTouchY(pointer: Int): Int {
        synchronized(this) { return touchY }
    }

    override fun getTouchEvents(): MutableList<Input.TouchEvent> {
        synchronized(this) {
            val len = touchEvents.size
            for (i in 0 until len) touchEventPool.free(touchEvents[i])
            touchEvents.clear()
            touchEvents.addAll(touchEventsBuffer)
            touchEventsBuffer.clear()
            return touchEvents
        }
    }

}